src/modules/audio/spatializers/oculus_spatializer.c

```diff
@@ -0,0 +1,225 @@
+    state.sources[idx].usedSourceThisPlayback = true;
+
+    float position[4], orientation[4];
+    lovrSourceGetPose(source, position, orientation);
+
+    ovrAudio_SetAudioSourcePos(state.context, idx, position[0], position[1], position[2]);
+
+    ovrAudio_SpatializeMonoSourceInterleaved(state.context, idx, &outStatus, output, input);
+
+    if (!lovrSourceIsPlaying(source)) { // Source is finished
+      state.sources[idx].source = NULL;
+      *spatializerMemo = -1;
+      if (outStatus & ovrAudioSpatializationStatus_Finished) { // Source done playing, echo tailoff is done
+        state.sources[idx].occupied = false;
+      }
+    }
+    return framesOut;
+  }
+  return 0;
+}
+
+static uint32_t oculus_spatializer_tail(float *scratch, float* output, uint32_t frames) {
+  bool didAnything = false;
+  for (int idx = 0; idx < state.sourceMax; idx++) {
+    if (state.sources[idx].occupied && !state.sources[idx].usedSourceThisPlayback) {
+      uint32_t outStatus = 0;
+      if (!didAnything) {
+        didAnything = true;
+        memset(output, 0, frames*sizeof(float)*2);
+      }
+      ovrAudio_SpatializeMonoSourceInterleaved(state.context, idx, &outStatus, scratch, NULL);
+      if (outStatus & ovrAudioSpatializationStatus_Finished) {
+        state.sources[idx].occupied = false;
+      }
+      for(int i = 0; i < frames*2; i++) {
+        output[i] += scratch[i];
+      }
+    }
+  }
+  return didAnything ? frames : 0;
+}
+
+// Oculus math primitives
+
+static void oculusUnpackQuat(ovrQuatf* oq, float* lq) {
+  oq->x = lq[0]; oq->y = lq[1]; oq->z = lq[2]; oq->w = lq[3];
+}
+static void oculusUnpackVec(ovrVector3f* ov, float *p) {
+  ov->x = p[0]; ov->y = p[1]; ov->z = p[2];
+}
+
+static void oculusRecreatePose(ovrPoseStatef *out, float position[4], float orientation[4]) {
+  ovrPosef pose;
+  oculusUnpackVec(&pose.Position, position);
+  oculusUnpackQuat(&pose.Orientation, orientation);
+  out->ThePose = pose;
+  float zero[4] = {0}; // TODO
+  oculusUnpackVec(&out->AngularVelocity, zero);
+  oculusUnpackVec(&out->LinearVelocity, zero);
+  oculusUnpackVec(&out->AngularAcceleration, zero);
+  oculusUnpackVec(&out->LinearAcceleration, zero);
+  out->TimeInSeconds = 0; //TODO-OS
+}
+
+static void oculus_spatializer_setListenerPose(float position[4], float orientation[4]) {
+  ovrPoseStatef pose;
+
+  oculusRecreatePose(&pose, position, orientation);
+
+  ma_mutex_lock(&state.poseLock); // Do nothing inside lock but make a copy of the pose
+  memcpy(&state.pose, &pose, sizeof(state.pose));
+  state.poseUpdated = true;
+  ma_mutex_unlock(&state.poseLock);
+}
+static void oculus_spatializer_source_create(Source *source) {
+  intptr_t *spatializerMemo = lovrSourceGetSpatializerMemoField(source);
+  *spatializerMemo = -1;
+}
+static void oculus_spatializer_source_destroy(Source *source) {
+  intptr_t *spatializerMemo = lovrSourceGetSpatializerMemoField(source);
+  if (*spatializerMemo >= 0)
+    state.sources[*spatializerMemo].source = NULL;

```

Comment:
Note this line means we MUST delete sources THEN call lovrAudioDestroy (in that order) or this line will crash